Code/GraphMol/RascalMCES/RascalMCES.cpp

```diff
@@ -0,0 +1,1086 @@
+                  checkEquivalentsAllowed(*starter.d_mol2);
+  }
+  while (!parts.empty()) {
+    if (opts.timeout != -1) {
+      checkTimeout(startTime, opts, clique, maxCliques, numSteps);
+    }
+    auto part = parts.back();
+    bool goDeeper = false;
+    bool backtrack = false;
+    if (opts.allBestMCESs) {
+      if (clique.size() + part->numParts() < starter.d_lowerBound) {
+        backtrack = true;
+      }
+    } else {
+      if (clique.size() + part->numParts() <= starter.d_lowerBound) {
+        backtrack = true;
+      }
+    }
+    if (!backtrack) {
+      if (opts.allBestMCESs) {
+        goDeeper = clique.size() + part->upperBound() >= starter.d_lowerBound;
+      } else {
+        goDeeper = clique.size() + part->upperBound() > starter.d_lowerBound;
+      }
+      if (goDeeper) {
+        if (!part->isEmpty()) {
+          std::shared_ptr<PartitionSet> nextPart(new PartitionSet(*part));
+          clique.push_back(nextPart->popLastVertex());
+          if (clique.size() == 1 && canDoEquivs &&
+              equivalentRootAlreadyDone(clique.front(), starter.d_vtxPairs,
+                                        starter.d_equivBonds1,
+                                        starter.d_equivBonds2, rootClasses)) {
+            clique.pop_back();
+            backtrack = true;
+          } else {
+            nextPart->pruneVertices(clique.back());
+            updateMaxClique(clique, starter.d_deltaYPoss, opts, *starter.d_mol1,
+                            *starter.d_mol2, starter.d_vtxPairs, maxCliques,
+                            starter.d_lowerBound);
+            parts.push_back(nextPart);
+          }
+        } else {
+          backtrack = true;
+        }
+      } else {
+        backtrack = true;
+      }
+    }
+    if (backtrack || (!parts.empty() && parts.back()->isEmpty())) {
+      while (!parts.empty()) {
+        if (parts.back()->isEmpty()) {
+          parts.pop_back();
+          if (!clique.empty()) {
+            clique.pop_back();
+          }
+        } else {
+          parts.back()->popLastVertex();
+          if (!parts.back()->isEmpty()) {
+            break;
+          }
+        }
+      }
+    }
+    if (parts.empty()) {
+      break;
+    }
+  }
+}
+
+bool deltaYExchangePossible(const ROMol &mol1, const ROMol &mol2) {
+  // A Delta-y exchange is an incorrect match when a cyclopropyl ring (the
+  // delta) is matched to a C(C)(C) group (the y) because they both have
+  // isomorphic line graphs.  This checks to see if that's something we need to
+  // worry about for these molecules.
+  const static std::unique_ptr<ROMol> delta(SmartsToMol("C1CC1"));
+  const static std::unique_ptr<ROMol> y(SmartsToMol("C(C)C"));
+  MatchVectType dontCare;
+  return (SubstructMatch(mol1, *delta, dontCare) &&
+          SubstructMatch(mol2, *y, dontCare)) ||
+         (SubstructMatch(mol2, *delta, dontCare) &&
+          SubstructMatch(mol1, *y, dontCare));
```

Comment:
your call as to which you prefer, but I find this a bit cleaner:
```suggestion
  auto hasSubstructMatch = [](const auto &m1, const auto &m2) -> bool {
    SubstructMatchParameters ps;
    ps.maxMatches = 1;
    return !SubstructMatch(m1,m2,ps).empty();
  };

  return (hasSubstructMatch(mol1, *delta) &&
          hasSubstructMatch(mol2, *y)) ||
         (hasSubstructMatch(mol2, *delta) &&
          hasSubstructMatch(mol1, *y));
```